router.go

```diff
@@ -36,23 +38,41 @@ type serviceProcedure struct {
 	procedure string
 }
 
+type serviceProcedureEncoding struct {
+	service   string
+	procedure string
+	encoding  transport.Encoding
+}
+
 // MapRouter is a Router that maintains a map of the registered
 // procedures.
 type MapRouter struct {
-	defaultService string
-	entries        map[serviceProcedure]transport.Procedure
+	defaultService            string
+	serviceProcedures         map[serviceProcedure]transport.Procedure
+	serviceProcedureEncodings map[serviceProcedureEncoding]transport.Procedure
+	supportedEncodings        map[serviceProcedure][]string
 }
 
 // NewMapRouter builds a new MapRouter that uses the given name as the
 // default service name.
 func NewMapRouter(defaultService string) MapRouter {
 	return MapRouter{
-		defaultService: defaultService,
-		entries:        make(map[serviceProcedure]transport.Procedure),
+		defaultService:            defaultService,
+		serviceProcedures:         make(map[serviceProcedure]transport.Procedure),
+		serviceProcedureEncodings: make(map[serviceProcedureEncoding]transport.Procedure),
+		supportedEncodings:        make(map[serviceProcedure][]string),
 	}
 }
 
 // Register registers the procedure with the MapRouter.
+// If the procedure does not specify its service name, the procedure will
+// inherit the default service name of the router.
+// Procedures should specify their encoding, and multiple procedures with the
+// same name and service name can exist if they handle different encodings.
+// However, specifying the encoding is optional since it was not required
+// in version 1.
```

Comment:
I mean, the contract at this point is that encoding is not required. This is essentially saying that "encoding isn't required because it wasn't required". We can simply say that specifying the encoding is recommended but optional (and so and so happens when you specify it and when you don't).